docs/gguf.md

```diff
@@ -0,0 +1,511 @@
+    ggml_type type;
+    // The offset of the tensor's data in this file in bytes.
+    // This offset is relative to `tensor_data`, not to the start
+    // of the file, to make it easier for writers to write the file.
+    // Readers should consider exposing this offset relative to the
+    // file to make it easier to read the data.
+    // Must be a multiple of `ALIGNMENT`.
+    uint64_t offset;
+};
+
+struct gguf_file_t {
+    // The header of the file.
+    gguf_header_t header;
+
+    // Tensor infos, which can be used to locate the tensor data.
+    gguf_tensor_info_t tensor_infos[header.tensor_count];
+
+    // Padding to the nearest multiple of `ALIGNMENT`.
+    uint8_t _padding[ALIGNMENT - (sizeof(header + tensor_infos) % ALIGNMENT)];
+
+    // Tensor data.
+    //
+    // This is arbitrary binary data corresponding to the weights of the model. This data should be close
+    // or identical to the data in the original model file, but may be different due to quantization or
+    // other optimizations for inference. Any such deviations should be recorded in the metadata or as
+    // part of the architecture definition.
+    //
+    // Each tensor's data must be stored within this array, and located through its `tensor_infos` entry.
+    // The offset of each tensor's data must be a multiple of `ALIGNMENT`, and the space between tensors
+    // should be padded to `ALIGNMENT` bytes.
+    uint8_t tensor_data[];
+};
+```
+
+## Standardized key-value pairs
+
+The following key-value pairs are standardized. This list may grow in the future as more use cases are discovered. Where possible, names are shared with the original model definitions to make it easier to map between the two.
+
+Not all of these are required, but they are all recommended. Keys that are required are bolded. For omitted pairs, the reader should assume that the value is unknown and either default or error as appropriate.
+
+The community can develop their own key-value pairs to carry additional data. However, these should be namespaced with the relevant community name to avoid collisions. For example, the `rustformers` community might use `rustformers.` as a prefix for all of their keys.
+
+If a particular community key is widely used, it may be promoted to a standardized key.
+
+### General
+
+#### Required
+
+- **`general.architecture: string`**: describes what architecture this model implements. All lowercase ASCII, with only `[a-z0-9]+` characters allowed. Known values include:
+  - `llama`
+  - `mpt`
+  - `gptneox`
+  - `gptj`
+  - `gpt2`
+  - `bloom`
+  - `falcon`
+  - `rwkv`
+- **`general.quantization_version: uint32`**: version of quantization scheme. Not required if the model is not quantized (i.e. no tensors are quantized). If any tensors are quantized, this _must_ be present.
+- **`general.alignment: uint32`**: the global alignment to use, as described above. This can vary to allow for different alignment schemes, but it must be a multiple of 8.
```

Comment:
Should this be required or default to 32 if not present as it is currently done in the implementation?

Reply:
I would suggest required to be a power of two and at least 32.

Reply:
32 for the tensor data - for the structs 8 should be fine.

Reply:
Maybe even better to have this in the header instead of a kv value?

Reply:
If in the header the tensor loader wont have to deal with the kv's

Reply:
Maybe I am missing something, but if this is the alignment of the structs, then it needs to be known before the structs are read, which includes the kv values, right?

Reply:
The alignment is only for tensor data.

Reply:
I suggest removing the alignment from kv and adding this to the header:
`uint32_t tensor_data_alignment;`

Reply:
Another option could be to remove the configurable alignment entirely and just align everything to a large value like 128. That should be enough for any architecture, and it would simplify the implementation.